src/base/bittorrent/peer_blacklist.hpp

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <regex>
+
+#include <libtorrent/torrent_info.hpp>
+
+#include "base/net/geoipmanager.h"
+
+#include "peer_filter_plugin.hpp"
+#include "peer_logger.hpp"
+
+// bad peer filter
+bool is_bad_peer(const lt::peer_info& info)
+{
+  std::string pid = info.pid.to_string().substr(0, 8);
+  std::regex id_filter("-(XL|SD|XF|QD|BN|DL)(\\d+)-");
+  std::regex ua_filter(R"((\d+.\d+.\d+.\d+|cacao_torrent))");
+  return std::regex_match(pid, id_filter) || std::regex_match(info.client, ua_filter);
+}
+
+// Unknown Peer filter
+bool is_unknown_peer(const lt::peer_info& info)
+{
+  QString country = Net::GeoIPManager::instance()->lookup(QHostAddress(info.ip.data()));
+  return info.client.find("Unknown") != std::string::npos && country == QLatin1String("CN");
+}
+
+// Offline Downloader filter
+bool is_offline_downloader(const lt::peer_info& info)
+{
+  unsigned short port = info.ip.port();
+  QString country = Net::GeoIPManager::instance()->lookup(QHostAddress(info.ip.data()));
+  return port >= 65000 && country == QLatin1String("CN") && info.client.find("Transmission") != std::string::npos;
+}
+
+// BitTorrent Media Player Peer filter
+bool is_bittorrent_media_player(const lt::peer_info& info)
+{
+  std::string pid = info.pid.to_string().substr(0, 8);
+  std::regex player_filter("-(UW\\w{4})-");
+  return !!std::regex_match(pid, player_filter);
+}
+
+
+// drop connection action
+void drop_connection(lt::peer_connection_handle ph)
+{
+  ph.disconnect(boost::asio::error::connection_refused, lt::operation_t::bittorrent, 0);
```

Comment:
Last parameter `disconnect_severity_t` is optional, so you can remove it.
```
void disconnect (error_code const& ec, operation_t op
    , disconnect_severity_t = peer_connection_interface::normal);
```

Reply:
unfortunately not... I even can't use constants... using constant (or dropping it and using default value) leads to linker error:
```
undefined reference to `libtorrent::peer_connection_interface::normal'
```
this happens at least on Debian stable and Ubuntu 20.04, and only when such constants are used in header files, everything is when they are used in .cpp. and this behavior is compile-specific, this issue is not reproducible (code compiles fine) on macOS (where I initially wrote this code) and Debian unstable).
building qBittorrent in C++17 mode also fixes that issue: https://stackoverflow.com/questions/40690260/undefined-reference-error-for-static-constexpr-member
so, that '0' at the end is workaround to guarantee successful build on almost any compiler

Reply:
I can't reproduce (`libtorrent::peer_connection_interface::normal'`) on MSVC, but MSVC don't like `0` for workaround, even change to C++17 mode.
```
Error	C2664	'void libtorrent::peer_connection_handle::disconnect(const libtorrent::error_code &,libtorrent::operation_t,libtorrent::disconnect_severity_t)': cannot convert argument 3 from 'int' to 'libtorrent::disconnect_severity_t' (compiling source file E:\qBittorrent\qBittorrent430\src\base\bittorrent\session.cpp)	qbt_base	E:\qBittorrent\qBittorrent430\src\base\bittorrent\peer_blacklist.hpp	48	
```

No problem if using jom for building, weird...

How about use `lt::disconnect_severity_t{0}` for workaround?
It is fine for MSVC and ubuntu 20.04.

Reply:
done, GCC10 on Debian unstable is happy with it too, thanks!